src/rollback_to_stable/rts.c

```diff
@@ -105,6 +105,124 @@ __wt_rts_progress_msg(WT_SESSION_IMPL *session, WT_TIMER *rollback_start, uint64
     }
 }
 
+/*
+ * __wt_rts_thread_chk --
+ *     Check to decide if the RTS thread should continue running.
+ */
+bool
+__wt_rts_thread_chk(WT_SESSION_IMPL *session)
+{
+    return (F_ISSET(S2C(session), WT_CONN_RTS_THREAD_RUN));
+}
+
+/*
+ * __wt_rts_thread_run --
+ *     Entry function for an RTS thread. This is called repeatedly from the thread group code so it
+ *     does not need to loop itself.
+ */
+int
+__wt_rts_thread_run(WT_SESSION_IMPL *session, WT_THREAD *thread)
+{
+    WT_DECL_RET;
+    WT_RTS_WORK_UNIT *entry;
+
+    WT_UNUSED(thread);
+
+    /* Wait here. */
+    if (F_ISSET(S2C(session), WT_CONN_RTS_THREAD_RUN))
+        __wt_cond_wait(session, S2C(session)->rts->thread_group.wait_cond, 10 * WT_THOUSAND, NULL);
+
+    /* Mark the session as an eviction thread session. */
+    F_SET(session, WT_SESSION_ROLLBACK_TO_STABLE);
+
+    while (!TAILQ_EMPTY(&S2C(session)->rts->rtsqh)) {
+        __wt_rts_pop_work(session, &entry);
+        if (entry == NULL)
+            break;
+
+        WT_ERR(__wt_rts_btree_work_unit(session, entry));
+    }
+
+    if (0) {
+err:
+        WT_RET_PANIC(session, ret, "rts thread error");
+    }
+    return (ret);
+}
+
+/*
+ * __wt_rts_thread_stop --
+ *     Shutdown function for an RTS thread.
+ */
+int
+__wt_rts_thread_stop(WT_SESSION_IMPL *session, WT_THREAD *thread)
+{
+    WT_UNUSED(thread);
+
+    /* Clear the eviction thread session flag. */
+    F_CLR(session, WT_SESSION_ROLLBACK_TO_STABLE);
+    return (0);
+}
+
+/*
+ * __wt_rts_thread_create --
+ *     Start RTS threads.
+ */
+static int
+__wt_rts_thread_create(WT_SESSION_IMPL *session)
+{
+    WT_CONNECTION_IMPL *conn;
+    uint32_t session_flags;
+
+    conn = S2C(session);
+
+    /* Set first, the thread might run before we finish up. */
+    F_SET(conn, WT_CONN_RTS_THREAD_RUN);
```

Comment:
should we use a write barrier here?

Reply:
I don't think that we need a barrier; even if it is executed out of order, there shouldn't be a problem.

Reply:
Then the comment is misleading. Set first sounds that this has to be written before anything else.

Reply:
Yes, there is a similar logic for the eviction threads as well. We set it up without a write barrier and started the thread group. Is the function call working as a barrier?

Reply:
Does it need only be set before the thread group starts? If that's the case, no barrier is needed because the system call to start the thread acts as a full barrier.